Features/#214 landuse areas

IlkaCu opened the pull request:
Fixes #214

gplssm reviewed (changes requested):
Please see the one comment in a changed file.

You could consider to add a bit more documentation in the SQL code. Just consider the case someone else would have to change anything ;-)
Maybe I'm a bit picky about this after reviewing and partially reimplementing old open_ego code about MV grid districts...

Apart from that, I must say that I get an error which I don't understand:
```
[2021-04-21 08:12:23,940] {logging_mixin.py:112} INFO - Running <TaskInstance: egon-data-processing-pipeline.extract-osm_landuse 2021-04-21T06:03:44.339975+00:00 [running]> on host RLI-NB-49
[2021-04-21 08:12:23,964] {postgres_operator.py:62} INFO - Executing: /*
OSM landuse sectors
Extract landuse areas from OpenStreetMap.
Cut the landuse with German boders (vg250) and make valid geometries.
Divide into 4 landuse sectors:
1. Residential
2. Retail
3. Industrial
4. Agricultural
__copyright__   = "Reiner Lemoine Institut"
__license__     = "GNU Affero General Public License Version 3 (AGPL-3.0)"
__url__         = "https://github.com/openego/data_processing/blob/master/LICENSE"
__author__      = "Ludee, IlkaCu"
*/


---------- ---------- ----------
-- Filter OSM Urban Landuse
---------- ---------- ----------

DELETE FROM openstreetmap.osm_polygon_urban; 

-- filter urban

INSERT INTO            openstreetmap.osm_polygon_urban
    SELECT  osm.gid ::integer AS gid,
            osm.osm_id ::integer AS osm_id,
            --osm.landuse ::text AS landuse,
            --osm.man_made ::text AS man_made,
            --osm.aeroway ::text AS aeroway,
            osm.name ::text AS name,
            --osm.way_area ::double precision AS way_area,
            '0' ::integer AS sector,
            'undefined'::text AS sector_name, 
            ST_AREA(ST_TRANSFORM(osm.geom,3035))/10000 ::double precision AS area_ha,
            osm.tags ::hstore AS tags,
            'outside' ::text AS vg250,
            ST_MULTI(ST_TRANSFORM(osm.geom,3035)) ::geometry(MultiPolygon,3035) AS geom
    FROM    openstreetmap.osm_polygon AS osm
    WHERE
        tags @> '"landuse"=>"residential"'::hstore OR 
        tags @> '"landuse"=>"commercial"'::hstore OR 
        tags @> '"landuse"=>"retail"'::hstore OR 
        tags @> '"landuse"=>"industrial;retail"'::hstore OR 

        tags @> '"landuse"=>"industrial"'::hstore OR 
        tags @> '"landuse"=>"port"'::hstore OR 
        tags @> '"man_made"=>"wastewater_plant"'::hstore OR
        tags @> '"aeroway"=>"terminal"'::hstore OR 
        tags @> '"aeroway"=>"gate"'::hstore OR 
        tags @> '"man_made"=>"works"'::hstore OR 

        tags @> '"landuse"=>"farmyard"'::hstore OR 
        tags @> '"landuse"=>"greenhouse_horticulture"'::hstore 
    ORDER BY    osm.gid;


-- OSM Urban Landuse Inside vg250

-- Calculate 'inside' vg250
UPDATE 	openstreetmap.osm_polygon_urban AS t1
	SET  	vg250 = t2.vg250
	FROM    (
		SELECT	osm.gid AS gid,
			'inside' ::text AS vg250
		FROM	boundaries.vg250_sta_union AS vg,
			openstreetmap.osm_polygon_urban AS osm
		WHERE  	vg.geometry && osm.geom AND
			ST_CONTAINS(vg.geometry,osm.geom)
		) AS t2
	WHERE  	t1.gid = t2.gid;

-- Calculate 'crossing' vg250
UPDATE 	openstreetmap.osm_polygon_urban AS t1
	SET  	vg250 = t2.vg250
	FROM    (
		SELECT	osm.gid AS gid,
			'crossing' ::text AS vg250
		FROM	boundaries.vg250_sta_union AS vg,
			openstreetmap.osm_polygon_urban AS osm
		WHERE  	osm.vg250 = 'outside' AND
			vg.geometry && osm.geom AND
			ST_Overlaps(vg.geometry,osm.geom)
		) AS t2
	WHERE  	t1.gid = t2.gid;


-- OSM outside of vg250

-- OSM error vg250
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_error_geom_vg250 CASCADE;
CREATE MATERIALIZED VIEW		openstreetmap.osm_polygon_urban_error_geom_vg250 AS
	SELECT	osm.*
	FROM	openstreetmap.osm_polygon_urban AS osm
	WHERE	osm.vg250 = 'outside' OR osm.vg250 = 'crossing';

-- index (id)
CREATE UNIQUE INDEX  	osm_polygon_urban_error_geom_vg250_gid_idx
		ON	openstreetmap.osm_polygon_urban_error_geom_vg250 (gid);

-- index GIST (geom)
CREATE INDEX  	osm_polygon_urban_error_geom_vg250_geom_idx
	ON	openstreetmap.osm_polygon_urban_error_geom_vg250
	USING	GIST (geom);

-- Sequence
DROP SEQUENCE IF EXISTS 	openstreetmap.osm_polygon_urban_vg250_cut_id CASCADE;
CREATE SEQUENCE 		openstreetmap.osm_polygon_urban_vg250_cut_id;

-- Cutting 'crossing' with vg250
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_vg250_cut;
CREATE MATERIALIZED VIEW		openstreetmap.osm_polygon_urban_vg250_cut AS
	SELECT	nextval('openstreetmap.osm_polygon_urban_vg250_cut_id') ::integer AS id,
		cut.gid ::integer AS gid,
		cut.osm_id ::integer AS osm_id,
		--cut.landuse ::text AS landuse,
		--cut.man_made ::text AS man_made,
		--cut.aeroway ::text AS aeroway,
		cut.name ::text AS name,
		--cut.way_area ::double precision AS way_area,
		cut.sector ::integer AS sector,
		cut.sector_name::text AS sector_name,
		cut.area_ha ::double precision AS area_ha,
		cut.tags ::hstore AS tags,
		cut.vg250 ::text AS vg250,
		GeometryType(cut.geom_new) ::text AS geom_type,
		ST_MULTI(ST_TRANSFORM(cut.geom_new,3035)) ::geometry(MultiPolygon,3035) AS geom
	FROM	(SELECT	poly.*,
			ST_INTERSECTION(poly.geom, cut.geometry) AS geom_new
		FROM	openstreetmap.osm_polygon_urban_error_geom_vg250 AS poly,
			boundaries.vg250_sta_union AS cut
		WHERE	poly.vg250 = 'crossing'
		) AS cut
	ORDER BY 	cut.gid;

-- index (id)
CREATE UNIQUE INDEX  	osm_polygon_urban_vg250_cut_gid_idx
		ON	openstreetmap.osm_polygon_urban_vg250_cut (gid);

-- index GIST (geom)
CREATE INDEX  	osm_polygon_urban_vg250_cut_geom_idx
	ON	openstreetmap.osm_polygon_urban_vg250_cut
	USING	GIST (geom);

-- 'crossing' Polygon to MultiPolygon
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_vg250_clean_cut_multi;
CREATE MATERIALIZED VIEW		openstreetmap.osm_polygon_urban_vg250_clean_cut_multi AS
	SELECT	nextval('openstreetmap.osm_polygon_gid_seq'::regclass) ::integer AS gid,
		cut.osm_id ::integer AS osm_id,
		--cut.landuse ::text AS landuse,
		--cut.man_made ::text AS man_made,
		--cut.aeroway ::text AS aeroway,
		cut.name ::text AS name,
		cut.sector ::integer AS sector,
		cut.sector_name::text AS sector_name,
		--cut.way_area ::double precision AS way_area,
		cut.area_ha ::double precision AS area_ha,
		cut.tags ::hstore AS tags,
		cut.vg250 ::text AS vg250,
		ST_MULTI(cut.geom) ::geometry(MultiPolygon,3035) AS geom
	FROM	openstreetmap.osm_polygon_urban_vg250_cut AS cut
	WHERE	cut.geom_type = 'POLYGON';

-- index (id)
CREATE UNIQUE INDEX  	osm_polygon_urban_vg250_clean_cut_multi_gid_idx
		ON	openstreetmap.osm_polygon_urban_vg250_clean_cut_multi (gid);

-- index GIST (geom)
CREATE INDEX  	osm_polygon_urban_vg250_clean_cut_multi_geom_idx
	ON	openstreetmap.osm_polygon_urban_vg250_clean_cut_multi
	USING	GIST (geom);

---------- ---------- ----------

-- clean cut
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_vg250_clean_cut;
CREATE MATERIALIZED VIEW		openstreetmap.osm_polygon_urban_vg250_clean_cut AS
	SELECT	nextval('openstreetmap.osm_polygon_gid_seq'::regclass) ::integer AS gid,
		cut.osm_id ::integer AS osm_id,
		--cut.landuse ::text AS landuse,
		--cut.man_made ::text AS man_made,
		--cut.aeroway ::text AS aeroway,
		cut.name ::text AS name,
		cut.sector ::integer AS sector,
		cut.sector_name::text AS sector_name, 
		--cut.way_area ::double precision AS way_area,
		cut.area_ha ::double precision AS area_ha,
		cut.tags ::hstore AS tags,
		cut.vg250 ::text AS vg250,
		cut.geom ::geometry(MultiPolygon,3035) AS geom
	FROM	openstreetmap.osm_polygon_urban_vg250_cut AS cut
	WHERE	cut.geom_type = 'MULTIPOLYGON';

-- index (id)
CREATE UNIQUE INDEX  	osm_polygon_urban_vg250_clean_cut_gid_idx
		ON	openstreetmap.osm_polygon_urban_vg250_clean_cut (gid);


-- remove 'outside' vg250
DELETE FROM	openstreetmap.osm_polygon_urban AS osm
	WHERE	osm.vg250 = 'outside';

-- remove 'outside' vg250
DELETE FROM	openstreetmap.osm_polygon_urban AS osm
	WHERE	osm.vg250 = 'crossing';

-- insert cut
INSERT INTO	openstreetmap.osm_polygon_urban
	SELECT	clean.*
	FROM	openstreetmap.osm_polygon_urban_vg250_clean_cut AS clean
	ORDER BY 	clean.gid;

-- insert cut multi
INSERT INTO	openstreetmap.osm_polygon_urban
	SELECT	clean.*
	FROM	openstreetmap.osm_polygon_urban_vg250_clean_cut_multi AS clean
	ORDER BY 	clean.gid;


---------- ---------- ----------
-- Update Sector Information 
---------- ---------- ----------

-- Sector 1. Residential
-- update sector
UPDATE 	openstreetmap.osm_polygon_urban
SET  	sector = '1', 
	sector_name = 'residential'
WHERE	tags @> '"landuse"=>"residential"'::hstore;

-- Sector 2. Retail
-- update sector
UPDATE 	openstreetmap.osm_polygon_urban
SET  	sector = '2', 
	sector_name = 'retail'
WHERE	tags @> '"landuse"=>"commercial"'::hstore OR 
		tags @> '"landuse"=>"retail"'::hstore OR 
		tags @> '"landuse"=>"industrial;retail"'::hstore;
	
-- Sector 3. Industrial
-- update sector
UPDATE 	openstreetmap.osm_polygon_urban
SET  	sector = '3', 
	sector_name = 'industrial'
WHERE	tags @> '"landuse"=>"industrial"'::hstore OR 
		tags @> '"landuse"=>"port"'::hstore OR 
		tags @> '"man_made"=>"wastewater_plant"'::hstore OR
		tags @> '"aeroway"=>"terminal"'::hstore OR 
		tags @> '"aeroway"=>"gate"'::hstore OR 
		tags @> '"man_made"=>"works"'::hstore;
	
-- Sector 4. Agricultural
-- update sector
UPDATE 	openstreetmap.osm_polygon_urban
	SET  	sector = '4', 
		sector_name = 'agricultural'
	WHERE	tags @> '"landuse"=>"farmyard"'::hstore OR 
		tags @> '"landuse"=>"greenhouse_horticulture"'::hstore;

-- Drop MViews which are not of special interest for downstream tasks 	
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_error_geom_vg250 CASCADE;
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_vg250_clean_cut_multi CASCADE;
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_vg250_cut CASCADE;

[2021-04-21 08:12:23,972] {base_hook.py:89} INFO - Using connection to: id: egon_data. Host: 127.0.0.1, Port: 59734, Schema: egon-data-sh, Login: egon, Password: XXXXXXXX, extra: None
[2021-04-21 08:12:23,975] {dbapi_hook.py:176} INFO - /*
OSM landuse sectors
Extract landuse areas from OpenStreetMap.
Cut the landuse with German boders (vg250) and make valid geometries.
Divide into 4 landuse sectors:
1. Residential
2. Retail
3. Industrial
4. Agricultural
__copyright__   = "Reiner Lemoine Institut"
__license__     = "GNU Affero General Public License Version 3 (AGPL-3.0)"
__url__         = "https://github.com/openego/data_processing/blob/master/LICENSE"
__author__      = "Ludee, IlkaCu"
*/


---------- ---------- ----------
-- Filter OSM Urban Landuse
---------- ---------- ----------

DELETE FROM openstreetmap.osm_polygon_urban; 

-- filter urban

INSERT INTO            openstreetmap.osm_polygon_urban
    SELECT  osm.gid ::integer AS gid,
            osm.osm_id ::integer AS osm_id,
            --osm.landuse ::text AS landuse,
            --osm.man_made ::text AS man_made,
            --osm.aeroway ::text AS aeroway,
            osm.name ::text AS name,
            --osm.way_area ::double precision AS way_area,
            '0' ::integer AS sector,
            'undefined'::text AS sector_name, 
            ST_AREA(ST_TRANSFORM(osm.geom,3035))/10000 ::double precision AS area_ha,
            osm.tags ::hstore AS tags,
            'outside' ::text AS vg250,
            ST_MULTI(ST_TRANSFORM(osm.geom,3035)) ::geometry(MultiPolygon,3035) AS geom
    FROM    openstreetmap.osm_polygon AS osm
    WHERE
        tags @> '"landuse"=>"residential"'::hstore OR 
        tags @> '"landuse"=>"commercial"'::hstore OR 
        tags @> '"landuse"=>"retail"'::hstore OR 
        tags @> '"landuse"=>"industrial;retail"'::hstore OR 

        tags @> '"landuse"=>"industrial"'::hstore OR 
        tags @> '"landuse"=>"port"'::hstore OR 
        tags @> '"man_made"=>"wastewater_plant"'::hstore OR
        tags @> '"aeroway"=>"terminal"'::hstore OR 
        tags @> '"aeroway"=>"gate"'::hstore OR 
        tags @> '"man_made"=>"works"'::hstore OR 

        tags @> '"landuse"=>"farmyard"'::hstore OR 
        tags @> '"landuse"=>"greenhouse_horticulture"'::hstore 
    ORDER BY    osm.gid;


-- OSM Urban Landuse Inside vg250

-- Calculate 'inside' vg250
UPDATE 	openstreetmap.osm_polygon_urban AS t1
	SET  	vg250 = t2.vg250
	FROM    (
		SELECT	osm.gid AS gid,
			'inside' ::text AS vg250
		FROM	boundaries.vg250_sta_union AS vg,
			openstreetmap.osm_polygon_urban AS osm
		WHERE  	vg.geometry && osm.geom AND
			ST_CONTAINS(vg.geometry,osm.geom)
		) AS t2
	WHERE  	t1.gid = t2.gid;

-- Calculate 'crossing' vg250
UPDATE 	openstreetmap.osm_polygon_urban AS t1
	SET  	vg250 = t2.vg250
	FROM    (
		SELECT	osm.gid AS gid,
			'crossing' ::text AS vg250
		FROM	boundaries.vg250_sta_union AS vg,
			openstreetmap.osm_polygon_urban AS osm
		WHERE  	osm.vg250 = 'outside' AND
			vg.geometry && osm.geom AND
			ST_Overlaps(vg.geometry,osm.geom)
		) AS t2
	WHERE  	t1.gid = t2.gid;


-- OSM outside of vg250

-- OSM error vg250
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_error_geom_vg250 CASCADE;
CREATE MATERIALIZED VIEW		openstreetmap.osm_polygon_urban_error_geom_vg250 AS
	SELECT	osm.*
	FROM	openstreetmap.osm_polygon_urban AS osm
	WHERE	osm.vg250 = 'outside' OR osm.vg250 = 'crossing';

-- index (id)
CREATE UNIQUE INDEX  	osm_polygon_urban_error_geom_vg250_gid_idx
		ON	openstreetmap.osm_polygon_urban_error_geom_vg250 (gid);

-- index GIST (geom)
CREATE INDEX  	osm_polygon_urban_error_geom_vg250_geom_idx
	ON	openstreetmap.osm_polygon_urban_error_geom_vg250
	USING	GIST (geom);

-- Sequence
DROP SEQUENCE IF EXISTS 	openstreetmap.osm_polygon_urban_vg250_cut_id CASCADE;
CREATE SEQUENCE 		openstreetmap.osm_polygon_urban_vg250_cut_id;

-- Cutting 'crossing' with vg250
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_vg250_cut;
CREATE MATERIALIZED VIEW		openstreetmap.osm_polygon_urban_vg250_cut AS
	SELECT	nextval('openstreetmap.osm_polygon_urban_vg250_cut_id') ::integer AS id,
		cut.gid ::integer AS gid,
		cut.osm_id ::integer AS osm_id,
		--cut.landuse ::text AS landuse,
		--cut.man_made ::text AS man_made,
		--cut.aeroway ::text AS aeroway,
		cut.name ::text AS name,
		--cut.way_area ::double precision AS way_area,
		cut.sector ::integer AS sector,
		cut.sector_name::text AS sector_name,
		cut.area_ha ::double precision AS area_ha,
		cut.tags ::hstore AS tags,
		cut.vg250 ::text AS vg250,
		GeometryType(cut.geom_new) ::text AS geom_type,
		ST_MULTI(ST_TRANSFORM(cut.geom_new,3035)) ::geometry(MultiPolygon,3035) AS geom
	FROM	(SELECT	poly.*,
			ST_INTERSECTION(poly.geom, cut.geometry) AS geom_new
		FROM	openstreetmap.osm_polygon_urban_error_geom_vg250 AS poly,
			boundaries.vg250_sta_union AS cut
		WHERE	poly.vg250 = 'crossing'
		) AS cut
	ORDER BY 	cut.gid;

-- index (id)
CREATE UNIQUE INDEX  	osm_polygon_urban_vg250_cut_gid_idx
		ON	openstreetmap.osm_polygon_urban_vg250_cut (gid);

-- index GIST (geom)
CREATE INDEX  	osm_polygon_urban_vg250_cut_geom_idx
	ON	openstreetmap.osm_polygon_urban_vg250_cut
	USING	GIST (geom);

-- 'crossing' Polygon to MultiPolygon
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_vg250_clean_cut_multi;
CREATE MATERIALIZED VIEW		openstreetmap.osm_polygon_urban_vg250_clean_cut_multi AS
	SELECT	nextval('openstreetmap.osm_polygon_gid_seq'::regclass) ::integer AS gid,
		cut.osm_id ::integer AS osm_id,
		--cut.landuse ::text AS landuse,
		--cut.man_made ::text AS man_made,
		--cut.aeroway ::text AS aeroway,
		cut.name ::text AS name,
		cut.sector ::integer AS sector,
		cut.sector_name::text AS sector_name,
		--cut.way_area ::double precision AS way_area,
		cut.area_ha ::double precision AS area_ha,
		cut.tags ::hstore AS tags,
		cut.vg250 ::text AS vg250,
		ST_MULTI(cut.geom) ::geometry(MultiPolygon,3035) AS geom
	FROM	openstreetmap.osm_polygon_urban_vg250_cut AS cut
	WHERE	cut.geom_type = 'POLYGON';

-- index (id)
CREATE UNIQUE INDEX  	osm_polygon_urban_vg250_clean_cut_multi_gid_idx
		ON	openstreetmap.osm_polygon_urban_vg250_clean_cut_multi (gid);

-- index GIST (geom)
CREATE INDEX  	osm_polygon_urban_vg250_clean_cut_multi_geom_idx
	ON	openstreetmap.osm_polygon_urban_vg250_clean_cut_multi
	USING	GIST (geom);

---------- ---------- ----------

-- clean cut
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_vg250_clean_cut;
CREATE MATERIALIZED VIEW		openstreetmap.osm_polygon_urban_vg250_clean_cut AS
	SELECT	nextval('openstreetmap.osm_polygon_gid_seq'::regclass) ::integer AS gid,
		cut.osm_id ::integer AS osm_id,
		--cut.landuse ::text AS landuse,
		--cut.man_made ::text AS man_made,
		--cut.aeroway ::text AS aeroway,
		cut.name ::text AS name,
		cut.sector ::integer AS sector,
		cut.sector_name::text AS sector_name, 
		--cut.way_area ::double precision AS way_area,
		cut.area_ha ::double precision AS area_ha,
		cut.tags ::hstore AS tags,
		cut.vg250 ::text AS vg250,
		cut.geom ::geometry(MultiPolygon,3035) AS geom
	FROM	openstreetmap.osm_polygon_urban_vg250_cut AS cut
	WHERE	cut.geom_type = 'MULTIPOLYGON';

-- index (id)
CREATE UNIQUE INDEX  	osm_polygon_urban_vg250_clean_cut_gid_idx
		ON	openstreetmap.osm_polygon_urban_vg250_clean_cut (gid);


-- remove 'outside' vg250
DELETE FROM	openstreetmap.osm_polygon_urban AS osm
	WHERE	osm.vg250 = 'outside';

-- remove 'outside' vg250
DELETE FROM	openstreetmap.osm_polygon_urban AS osm
	WHERE	osm.vg250 = 'crossing';

-- insert cut
INSERT INTO	openstreetmap.osm_polygon_urban
	SELECT	clean.*
	FROM	openstreetmap.osm_polygon_urban_vg250_clean_cut AS clean
	ORDER BY 	clean.gid;

-- insert cut multi
INSERT INTO	openstreetmap.osm_polygon_urban
	SELECT	clean.*
	FROM	openstreetmap.osm_polygon_urban_vg250_clean_cut_multi AS clean
	ORDER BY 	clean.gid;


---------- ---------- ----------
-- Update Sector Information 
---------- ---------- ----------

-- Sector 1. Residential
-- update sector
UPDATE 	openstreetmap.osm_polygon_urban
SET  	sector = '1', 
	sector_name = 'residential'
WHERE	tags @> '"landuse"=>"residential"'::hstore;

-- Sector 2. Retail
-- update sector
UPDATE 	openstreetmap.osm_polygon_urban
SET  	sector = '2', 
	sector_name = 'retail'
WHERE	tags @> '"landuse"=>"commercial"'::hstore OR 
		tags @> '"landuse"=>"retail"'::hstore OR 
		tags @> '"landuse"=>"industrial;retail"'::hstore;
	
-- Sector 3. Industrial
-- update sector
UPDATE 	openstreetmap.osm_polygon_urban
SET  	sector = '3', 
	sector_name = 'industrial'
WHERE	tags @> '"landuse"=>"industrial"'::hstore OR 
		tags @> '"landuse"=>"port"'::hstore OR 
		tags @> '"man_made"=>"wastewater_plant"'::hstore OR
		tags @> '"aeroway"=>"terminal"'::hstore OR 
		tags @> '"aeroway"=>"gate"'::hstore OR 
		tags @> '"man_made"=>"works"'::hstore;
	
-- Sector 4. Agricultural
-- update sector
UPDATE 	openstreetmap.osm_polygon_urban
	SET  	sector = '4', 
		sector_name = 'agricultural'
	WHERE	tags @> '"landuse"=>"farmyard"'::hstore OR 
		tags @> '"landuse"=>"greenhouse_horticulture"'::hstore;

-- Drop MViews which are not of special interest for downstream tasks 	
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_error_geom_vg250 CASCADE;
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_vg250_clean_cut_multi CASCADE;
DROP MATERIALIZED VIEW IF EXISTS	openstreetmap.osm_polygon_urban_vg250_cut CASCADE;

[2021-04-21 08:12:29,489] {taskinstance.py:1150} ERROR - index row size 2712 exceeds btree version 4 maximum 2704 for index "ix_openstreetmap_osm_polygon_urban_geom"
DETAIL:  Index row references tuple (4,7) in relation "osm_polygon_urban".
HINT:  Values larger than 1/3 of a buffer page cannot be indexed.
Consider a function index of an MD5 hash of the value, or use full text indexing.
Traceback (most recent call last):
  File "/home/guido/miniconda3/envs/eGon-data/lib/python3.7/site-packages/airflow/models/taskinstance.py", line 984, in _run_raw_task
    result = task_copy.execute(context=context)
  File "/home/guido/miniconda3/envs/eGon-data/lib/python3.7/site-packages/airflow/operators/postgres_operator.py", line 65, in execute
    self.hook.run(self.sql, self.autocommit, parameters=self.parameters)
  File "/home/guido/miniconda3/envs/eGon-data/lib/python3.7/site-packages/airflow/hooks/dbapi_hook.py", line 177, in run
    cur.execute(s)
psycopg2.errors.ProgramLimitExceeded: index row size 2712 exceeds btree version 4 maximum 2704 for index "ix_openstreetmap_osm_polygon_urban_geom"
DETAIL:  Index row references tuple (4,7) in relation "osm_polygon_urban".
HINT:  Values larger than 1/3 of a buffer page cannot be indexed.
Consider a function index of an MD5 hash of the value, or use full text indexing.

[2021-04-21 08:12:29,493] {taskinstance.py:1194} INFO - Marking task as FAILED. dag_id=egon-data-processing-pipeline, task_id=extract-osm_landuse, execution_date=20210421T060344, start_date=20210421T061223, end_date=20210421T061229
[2021-04-21 08:12:33,513] {local_task_job.py:102} INFO - Task exited with return code 1
```

IlkaCu commented on src/egon/data/processing/loadarea/osm_landuse_extraction.sql:
```diff
@@ -0,0 +1,264 @@
+
+INSERT INTO            openstreetmap.osm_polygon_urban
+    SELECT  osm.gid ::integer AS gid,
+            osm.osm_id ::integer AS osm_id,
```
Done!

gplssm replied:
:+1:

IlkaCu commented:
I replaced the btree indexing by a GiST index. I hope this solves the problem for you. 

I also tried to extend the SQL documentation at least a little bit and removed commented parts of the code.

gplssm commented:
Thanks for the changes! I was able to execute it and took a look on the data. Seems to be quite decent coverage!

One minor thing I would like to change: renaming the resulting table from osm_polygon_urban to either landuse or landuse_polygons. This table name would tell a bit more what's inside the data.
